src/agent/mod.rs

```diff
@@ -47,15 +47,36 @@ fn run_heartbeat(url: &str, token: &str) {
     });
 }
 
+fn try_with_report<T>(api: &AgentApi, res: Fallible<T>) -> Option<T> {
+    match res {
+        Err(err) => {
+            utils::report_failure(&err);
+            if let Err(e) = api
+                .report_error(format!("{}", err.find_root_cause()))
+                .with_context(|_| "error encountered")
+            {
+                utils::report_failure(&e);
+            }
+            None
+        }
+        Ok(res) => Some(res),
+    }
+}
+
 pub fn run(url: &str, token: &str, threads_count: usize, docker_env: &str) -> Fallible<()> {
     let agent = Agent::new(url, token)?;
     let db = results::ResultsUploader::new(&agent.api);
 
     run_heartbeat(url, token);
 
     loop {
-        let ex = agent.experiment()?;
-        crate::runner::run_ex(&ex, &db, threads_count, &agent.config, docker_env)?;
-        agent.api.complete_experiment()?;
+        try_with_report(&agent.api, agent.experiment())
+            .and_then(|ex| {
+                try_with_report(
+                    &agent.api,
+                    crate::runner::run_ex(&ex, &db, threads_count, &agent.config, docker_env),
+                )
+            })
+            .and_then(|()| try_with_report(&agent.api, agent.api.complete_experiment()));
```

Comment:
This could be refactored by putting the old content of the loop in a `run_experiment` function. That allows you to still use `?` in the loop code, and avoids the `try_with_report` function.

Reply:
Do you mean completely removing `try_with_report` or only the call with `api.complete_experiment()`?

Reply:
Something like:

```rust
loop {
    if let Err(err) = run_experiment(&...) {
        // Your code that reports the error
    }
}
```